Implementation of np.corrcoef

wyjw opened the pull request:
As in #70, an implementation of corrcoef.

There is a small issue during testing in the clipping of a matrix of all nan, which I have described in #1072.

yjwusf committed:
here

wyjw committed:
Implementation of np.corrcoef

wyjw committed:
Update lax_numpy_test.py

hawkinsp commented on jax/numpy/lax_numpy.py:
```diff
@@ -2625,6 +2625,30 @@ def cov(m, y=None, rowvar=True, bias=False, ddof=None, fweights=None,
 
+@_wraps(onp.corrcoef)
+def corrcoef(x, y=None, rowvar=True, bias=None, ddof=None):
+  msg = ("jax.numpy.cov not implemented for nontrivial {}. "
```
Two things:
a) this should probably say "corrcoef" not "cov".
b) can you please move the message into the `if` condition? No need to allocate a string in the non-error case. Alternatively, you could just omit the error completely, assuming that all we are missing is the corresponding code in `jax.numpy.cov`.

hawkinsp commented on jax/numpy/lax_numpy.py:
```diff
@@ -2625,6 +2625,30 @@ def cov(m, y=None, rowvar=True, bias=False, ddof=None, fweights=None,
+  if y is not None: raise NotImplementedError(msg.format('y'))
+
+  c = cov(x, y, rowvar)
+  try:
```
I'd prefer to check the number of dimensions explicitly rather than relying on an exception from `np.diag`.

hawkinsp commented on jax/numpy/lax_numpy.py:
```diff
@@ -2625,6 +2625,30 @@ def cov(m, y=None, rowvar=True, bias=False, ddof=None, fweights=None,
+      d = diag(c)
+  except ValueError:
+      # scalar
+      return divide(c, c)
```
This seems strange. Isn't this always 1 for finite values? Am I misreading it? If it's intentional, add a comment.

hawkinsp reviewed (commented):
There appear to be test failures on CPU. Do you think these are all the same bug as https://github.com/google/jax/issues/1072 ? As far as I can see that bug only reproduces on GPU, but the NaN disagreement here clearly is happening on CPU since that's the backend our Travis CI builder is using. Is that the case?

Is there a simple test case for bad `np.clip` semantics that reproduces on CPU too? (You can try CPU by setting the environment variable `CUDA_VISIBLE_DEVICES=0`.)

Either way, we might need to temporarily avoid generating NaNs in the test to check this in.

yjwusf committed:
Merge branch 'master' of https://github.com/google/jax into corrcoef

yjwusf committed:
made changes to corrcoef

yjwusf committed:
here

yjwusf committed:
corrcoef

wyjw committed:
Merge pull request #1 from wyjw/corrcoef

Corrcoef

wyjw commented:
I'm sorry I messed up something with my local branch. I'll try to fix it.

wyjw committed:
Revert "Corrcoef"

wyjw committed:
Merge pull request #2 from wyjw/revert-1-corrcoef

Revert "Corrcoef"

wyjw committed:
Accidental mistake when reverting.

wyjw committed:
Update pxla.py

wyjw committed:
Update pxla.py

wyjw committed:
Made changes based on review.

wyjw commented:
Once again I apologize for the spam of commits. I have fixed it and fixed the code with respect to the review.

As far as my attempt with np.clip and nan go, it still seems to give an error on cpu.

```
import os
os.environ["CUDA_VISIBLE_DEVICES"]="0"
import jax.numpy as np
a = np.full((3,3,), np.nan)
np.clip(a, -1, 1)
```

If it helps, the OS I am using is Ubuntu 18.04, Python version 3.6.8 and jax is updated to the latest version.

wyjw committed:
added shape check

hawkinsp commented:
What version of `jaxlib` do you have? For me, your test case prints:
```
DeviceArray([[nan, nan, nan],
             [nan, nan, nan],
             [nan, nan, nan]], dtype=float32)
```
which seems like the right thing (and matches numpy).

hawkinsp commented on jax/numpy/lax_numpy.py:
```diff
@@ -2625,6 +2625,25 @@ def cov(m, y=None, rowvar=True, bias=False, ddof=None, fweights=None,
+@_wraps(onp.corrcoef)
+def corrcoef(x, y=None, rowvar=True, bias=None, ddof=None):
+  c = cov(x, y, rowvar)
+  if isscalar(c) or (len(c_shape) != 1 and len(c_shape) != 2):
```
You don't define `c_shape`, but `len(shape(c)) == 0` would seem to cover it.

Note that JAX doesn't distinguish between 0D arrays and scalars.

wyjw committed:
shape_c variable taken out

wyjw commented:
I have jaxlib 0.1.22. On my machine, it prints a square matrix of -1s.

hawkinsp commented:
Oops! I had an important type before. This 
```
os.environ["CUDA_VISIBLE_DEVICES"]="0"
```
should be
```
os.environ["CUDA_VISIBLE_DEVICES"]=""
```

So I think that repro is still a GPU bug. It's likely also that setting
```
XLA_FLAGS="--xla_gpu_enable_fast_min_max=false"
```
would fix the problem.

I'm still wondering what's happening for the CPU case, though; that sounds like a different issue. (In general NaN semantics aren't heavily tested; for your immediate needs the right fix is to change the test case harness not to generate NaNs.)

wyjw committed:
Update lax_numpy_test.py

wyjw committed:
make test more robust

wyjw committed:
Update lax_numpy_test.py

wyjw commented:
The NaNs were generated by the standard deviation being 0 for some matrices, and calculating the coefficient was done by dividing by the standard deviation. I added an if-case in the test, and it should work now.

Thanks!

hawkinsp commented on jax/numpy/lax_numpy.py:
```diff
@@ -2625,6 +2625,25 @@ def cov(m, y=None, rowvar=True, bias=False, ddof=None, fweights=None,
+@_wraps(onp.corrcoef)
+def corrcoef(x, y=None, rowvar=True, bias=None, ddof=None):
+  c = cov(x, y, rowvar)
+  if isscalar(c) or (len(shape(c)) != 1 and len(shape(c)) != 2):
```
JAX doesn't distinguish between scalars and 0D arrays. Perhaps there is value in testing for both (e.g., if a numpy classic array is passed here).

Why the two length tests? Why not `len(shape(c) == 0`?

wyjw replied:
Fixed.

hawkinsp reviewed (commented):
Looks great! Thanks for the PR. One more tiny suggestion.

wyjw committed:
Update lax_numpy.py

hawkinsp commented:
Looks great! Thanks for the PR!
